### PR TITLE
Make entire redux state tree immutable

### LINKED ISSUE #238: Make whole state immutable using redux-immutable
We should use [redux-immutable](https://github.com/gajus/redux-immutable) to make the entire state immutable. This will require some refactoring of the selectors.

### PR DESCRIPTION
**Pull request by rehandalal**
Fixes #238.

r?

### CODE REVIEW

File: src/state/router/selectors.js [hunk 1]
```diff
@@ -59,39 +60,41 @@ export function getUrlParam(state, key, defaultsTo) {
...
 
-  return search
-    .slice(1)
-    .split('&')
-    .map(item => item.split('='))
-    .reduce((obj, [key, value]) => {
-      obj[key] = value === undefined ? true : value;
-      return obj;
-    }, {});
+  return fromJS(
```
Comment (magopian) [inline src/state/router/selectors.js]: I'm confused, isn't `search` taken from the state, which is fully immutable? Why the `fromJS`?
Also it feels weird to have a `fromJS` here, and in `getCurrentUrl` do a `toJS` on the returned value from here...

**Comment by rehandalal** [comment]
> Also I'm wondering about the session related files removal: was it a leftover? Was it moved to auth some time ago and wasn't properly cleaned up at that time?

Session was a bad name for what that was. It was leftover from the normandy UI transition and was used to power the "user history" sidebar that we no longer use. No code was referencing this part of the state and I didn't like the general implementation so I scrapped it.

File: src/state/router/selectors.js (hunk 1 above)
Comment (rehandalal) [reply src/state/router/selectors.js]: `search` is taken from the state. It is however a string so it's not immutable. We subsequently split it (transforming it to a standard `Array`) and perform some additional processing on it. The `fromJS` here converts the standard `Array` into an immutable `List`.

Comment (rehandalal) [reply src/state/router/selectors.js]: Perhaps i should add a comment explaining

Comment (magopian) [reply src/state/router/selectors.js]: ah, makes sense, thanks for the explanation. And yeah, would love a comment ;)

**Pull request merged**